### PR TITLE
Adds the --pulp-no-leftover-objs pytest arg

### PR DESCRIPTION
**Pull request by bmbouter**
[noissue]

### CODE REVIEW

**Comment by bmbouter** [comment]
With this a plugin could define something like this pulp_file example in their conftest.py

```
def pytest_check_for_leftover_pulp_objects(config):
    file_client = gen_file_client()

    types_to_check = [
        RepositoriesFileApi(file_client),
        # ContentFilesApi(file_client),  # It's ok to leave behind orphans
        RemotesFileApi(file_client),
        PublicationsFileApi(file_client),
        DistributionsFileApi(file_client),
    ]
    for type_to_check in types_to_check:
        if type_to_check.list().count > 0:
            raise Exception(f"This test left over a {type_to_check}.")
```

**Comment by bmbouter** [comment]
So this is neat and all, but I realized that pulpcore could check the generic list endpoints of the following object types which would not require plugins to add any code! I'm going to revise this to do that instead. At least I learned a lot about how to write a pytest plugin.

File: pulp_smash/pulp3/pytest_plugin/newhooks.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+def pytest_check_for_leftover_pulp_objects(config):
```
Comment (mdellweg) [inline pulp_smash/pulp3/pytest_plugin/newhooks.py:1]: Is this filename a common one, or taken from an example? Should we name it `smashhooks` or `pulphooks`?

File: pulp_smash/pulp3/pytest_plugin/__init__.py [hunk 2]
```diff
@@ -29,7 +30,52 @@
...
+    if item.config.getoption("--pulp-no-leftover-objs"):
+        item.config.hook.pytest_check_for_leftover_pulp_objects(config=item.config)
+
+
+## pytest configuration
+
+
 def pytest_configure(config):
+    if config.pluginmanager.hasplugin("xdist") and config.getoption("-n"):
+        raise Exception("The --pulp-no-leftover-objs cannot be used with -n from xdist")
```
Comment (mdellweg) [inline pulp_smash/pulp3/pytest_plugin/__init__.py:77]: This condition seems incomplete to me.

**Comment by bmbouter** [comment]
@mdellweg what name did you want for the argument? I forgot the suggestion from the video call.

**Comment by bmbouter** [comment]
[Here's an example](https://github.com/pulp/pulp_file/pull/681) of the hook being used by pulp_file.

**Comment by bmbouter** [comment]
@mdellweg this is ready for re-review and I'll rename the arg when you tell me what sounds best to you.

File: pulp_smash/pulp3/pytest_plugin/__init__.py [hunk 3]
```diff
@@ -29,7 +38,67 @@
...
+## Pulp No Leftover Objs Feature
+
+
+def pytest_addoption(parser):
+    group = parser.getgroup("pulp-smash")
+    group.addoption(
+        "--pulp-no-leftover-objs",
+        action="store_true",
+        dest="pulp_no_leftover_objs",
+        help="Enable this to have Pulp plugins check for objects leftover by tests.",
```
Comment (mdellweg) [inline pulp_smash/pulp3/pytest_plugin/__init__.py:50]:
```suggestion
        "--pulp-no-leftovers",
        action="store_true",
        dest="pulp_no_leftovers",
        help="Enable this to have Pulp plugins check for objects leftover by tests.",
```

I thought calling them objects is redundant. And saving few keystrokes will hopefully help people to run it more often.

File: pulp_smash/pulp3/pytest_plugin/__init__.py [hunk 4]
```diff
@@ -29,7 +38,67 @@
...
+
+    types_to_check = [
+        ContentguardsApi(pulpcore_client),
+        DistributionsApi(pulpcore_client),
+        PublicationsApi(pulpcore_client),
+        RepositoriesApi(pulpcore_client),
+    ]
+    for type_to_check in types_to_check:
+        if type_to_check.list().count > 0:
+            raise Exception(f"This test left over a {type_to_check}.")
```
Comment (mdellweg) [inline pulp_smash/pulp3/pytest_plugin/__init__.py:71]: is `repr(type_to_check)` readable? Also maybe call it `model_to_check`.

File: pulp_smash/pulp3/pytest_plugin/__init__.py (hunk 3 above)
Comment (bmbouter) [reply pulp_smash/pulp3/pytest_plugin/__init__.py:50]: I changed this in my next push.

File: pulp_smash/pulp3/pytest_plugin/__init__.py (hunk 4 above)
Comment (bmbouter) [reply pulp_smash/pulp3/pytest_plugin/__init__.py:71]: I switched this to model_to_check in my next push.

Comment (bmbouter) [reply pulp_smash/pulp3/pytest_plugin/__init__.py:71]: I think it's pretty readable it produces a statement like: `This test left over a <class 'pulpcore.client.pulpcore.api.distributions_api.DistributionsApi'>.`

**Review by mdellweg — APPROVED** [review]

**Pull request merged**